File: Algorithms/py-algo/BinarySearchTree.py
# balanced Binary Search Tree


class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class Tree:
    def __init__(self):
        self.root = None

    def insert(self, data):
        newNode = Node(data)

        if self.root == None:
            self.root = newNode
        else:
            self.insert_node(self.root, newNode)

    def insert_node(self, node, new_node):
        if new_node.data < node.data:
            if node.left == None:
                node.left = new_node
            else:
                self.insert_node(node.left, new_node)
        else:
            if node.right == None:
                node.right = new_node
            else:
                self.insert_node(node.right, new_node)

    def remove(self, data):
        self.root = self.remove_node(self.root, data)

    def remove_node(self, node, key):
        if node == None:
            return None
        
        elif key < node.data:
            node.left = self.remove_node(node.left, key)
            return node
        
        elif key > node.data:
            node.right = self.remove_node(node.right, key)
            return node
        
        else:
            if node.left == None and node.right == None:
                node = None
                return node
            
            if node.left == None:
                node = node.left
                return node
            
            elif node.right == None:
                node = node.right
                return node
            
            aux = self.find_min_node(node.right)
            node.data = aux.data

            node.right = self.remove_node(node.right, aux.data)
            return node

    def find_min_node(self,node):
        if node.left == None:
            return None

        else:
            return self.find_min_node(node.left)
        
    def get_root_node(self):
        return self.root
    

bst = Tree()

bst.insert(23)
bst.insert(22)
bst.insert(43)
bst.insert(56)
bst.insert(6)
bst.insert(3)

print(bst.get_root_node())

File: Algorithms/js-algo/BinarySearchTree.js
// balanced Binary Search Tree


class Node{
    constructor(data){
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class Tree{
    constructor(){
        this.root = null
    }

    insert(data){
        const newNode = new Node(data);

        if(this.root === null){
            this.root = newNode;
        } else {
            this.insertNode(this.root, newNode);
        }
    }

    insertNode(node, newNode){
        if(newNode.data < node.data){
            if(node.left === null){
                node.left = newNode;
            } else {
                this.insertNode(node.left, newNode);
            }
        } else {
            if(node.right === null){
                node.right = newNode
            } else {
                this.insertNode(node.right, newNode);
            }
        }
    }

    remove(data){
        this.root = this.removeNode(this.root, data);
    }

    removeNode(node, key){
        if(node === null){
            return null
        }

        else if(key < node.data){
            node.left = this.removeNode(node.left, key);
            return node;
        }

        else if (key > node.data){
            node.right = this.removeNode(node.right, key);
            return node;
        }

        else{
            if(node.right === null && node.left === null){
                node = null;
                return node

            }
            
            if(node.left === null) {
                node = node.left;
                return node;
            }

            else if (node.right === null){
                node = node.right;
                return node;
            }

            const aux = this.findMinNode(node.right);
            node.data = aux.data;

            node.right = this.removeNode(node.right, aux.data);
            return node;
        }
    }

    findMinNode(node){
        if(node.left === null){
            return node
        } else
            return this.findMinNode(node.left)
    }

    getRootNode(){
        return this.root
    }
}

const bst = new Tree();

bst.insert(25);
bst.insert(14);
bst.insert(35);
bst.insert(13);
bst.insert(33);
bst.insert(2);


console.log(bst.getRootNode());